(fix) `Go-to-top` button shouldn't be visible if user is on top of a page

coderatomy opened the pull request:
## Summary

User shouldn't see the `go-to-top` icon/button once on top of the page
Closes #628

## Changes

- Added `useIsInViewPort` hook

## Screenshots

coderatomy committed:
add useIsInViewPort hook

coderatomy commented:
In continuation of the discussion in issue #672, and delving deeper into this topic, @tuxology, it appears that relying solely on a ref to a specific element might be insufficient, as pointed out by @chaitanya-cloudknox.

Consider this scenario: if we attach a `ref` to an element at the top of the home page, and utilizing the `useIsInViewPort` hook, the button displays correctly when the user scrolls down and the element is within the viewport. However, if the user navigates to another page, like the profile page, the `ref` element is out of the viewport. Consequently, the button won't appear as intended

coderatomy commented on zubhub_frontend/zubhub/src/hooks/useIsInViewPort.hook.js:
```diff
@@ -0,0 +1,24 @@
+import { useState, useMemo, useEffect } from "react";
+
+
+export function useIsInViewport(ref) {
```
this hook generally watches over the reference element `ref` using the `IntersectionObserver` API. This will be used to check if an element is in the `view port`

chaitanya-cloudknox replied:
Can you move the IntersectionObserver creation to the useEffect as calling disconnect will stop emitting any new events?

Usually ref won't change, the value ref.current changes.  Just curious how the useEffect is working if the target element is remounted or re-rendered.

tuxology replied:
Any updates on this @coderatomy? This is a valid question

coderatomy replied:
Sorry for the late reply @tuxology @chaitanya-cloudknox 

Ideally, the useEffect hook handles remounting or re-rendering of the target element by cleaning up the previous IntersectionObserver and creating a new one to observe the updated element, thanks to the dependency array [ref]. This ensures that the hook stays in sync with the current state of the DOM.

tuxology commented:
@kamthamc Are you ok with this approach?

chaitanya-cloudknox commented:
> In continuation of the discussion in issue #672, and delving deeper into this topic, @tuxology, it appears that relying solely on a ref to a specific element might be insufficient, as pointed out by @chaitanya-cloudknox.
> 
> Consider this scenario: if we attach a `ref` to an element at the top of the home page, and utilizing the `useIsInViewPort` hook, the button displays correctly when the user scrolls down and the element is within the viewport. However, if the user navigates to another page, like the profile page, the `ref` element is out of the viewport. Consequently, the button won't appear as intended

I didn't understand it but the code looks fine. I have couple comments on it

coderatomy commented:
My main concern here @chaitanya-cloudknox is that using `ref` won't be the best solution here, just because we have different pages.
To get this best @chaitanya-cloudknox, where do you think the ref should be attached?

chaitanya-cloudknox commented:
Tell me more about it.
I am assuming it is for the go to top button. 
My understanding is in the useEffect you create a listener with the ref.current dependency (since the current might change with re render and ref might not) and during unmount you disconnect the listeners.

If the user navigates to different page we won't be showing the go to top button until user scrolls to the bottom. If we check the case if current is not null/undefined and skip the part of adding event listeners would solve the problem. And also this hook will be only invoked if its consumed in the page, i am assuming we use this hook in each page separately

coderatomy committed:
use window.pageYOffset instead

coderatomy commented:
> Tell me more about it. I am assuming it is for the go to top button. My understanding is in the useEffect you create a listener with the ref.current dependency (since the current might change with re render and ref might not) and during unmount you disconnect the listeners.
> 
> If the user navigates to different page we won't be showing the go to top button until user scrolls to the bottom. If we check the case if current is not null/undefined and skip the part of adding event listeners would solve the problem. And also this hook will be only invoked if its consumed in the page, i am assuming we use this hook in each page separately

I have changed to another approach. So basically, we wanted to show the button conditionally once the user is not at the top of the page. So adding the ref to the button itself is insufficient. I have updated to `window.pageYOffset`

Here I intend to show the button once there is any slight scroll down.
Simply, once a user is scrolling down, it is shown, otherwise not. I think this is easier and straight forward @chaitanya-cloudknox 

[Screencast from 2023-10-08 09-58-31.webm](https://github.com/unstructuredstudio/zubhub/assets/141822315/717554b1-1ba4-44f7-bf9e-5e4d7cf8164c)
